File: app/src/main/java/fr/free/nrw/commons/actions/ThanksClient.java
package fr.free.nrw.commons.actions;

import org.wikipedia.csrf.CsrfTokenClient;
import org.wikipedia.dataclient.Service;

import javax.inject.Inject;
import javax.inject.Named;
import javax.inject.Singleton;

import fr.free.nrw.commons.CommonsApplication;
import io.reactivex.Observable;

/**
 * Facilitates the Wkikimedia Thanks api extension, as described in the 
 * api documentation: "The Thanks extension includes an API for sending thanks"
 * 
 * In simple terms this class is used by a user to thank someone for adding 
 * contribution to the commons platform
 */
@Singleton
public class ThanksClient {

    private final CsrfTokenClient csrfTokenClient;
    private final Service service;

    @Inject
    public ThanksClient(@Named("commons-csrf") CsrfTokenClient csrfTokenClient,
                        @Named("commons-service") Service service) {
        this.csrfTokenClient = csrfTokenClient;
        this.service = service;
    }

    /**
     * Handles the Thanking logic
     * @param revisionID The revision ID you would like to thank someone for
     * @return if thanks was successfully sent to intended recipient, returned as a boolean observable
     */
    public Observable<Boolean> thank(long revisionId) {
        try {
            return service.thank(String.valueOf(revisionId), null,
                    csrfTokenClient.getTokenBlocking(),
                    CommonsApplication.getInstance().getUserAgent())
                    .map(mwQueryResponse -> mwQueryResponse.getSuccessVal() == 1);
        } catch (Throwable throwable) {
            return Observable.just(false);
        }
    }
}

File: app/src/main/java/fr/free/nrw/commons/actions/ThanksClient.kt
package fr.free.nrw.commons.actions

import fr.free.nrw.commons.CommonsApplication
import fr.free.nrw.commons.di.NetworkingModule.NAMED_COMMONS_CSRF
import io.reactivex.Observable
import fr.free.nrw.commons.auth.csrf.CsrfTokenClient
import fr.free.nrw.commons.auth.csrf.InvalidLoginTokenException
import fr.free.nrw.commons.auth.login.LoginFailedException
import javax.inject.Inject
import javax.inject.Named
import javax.inject.Singleton

/**
 * Client for the Wkikimedia Thanks API extension
 * Thanks are used by a user to show gratitude to another user for their contributions
 */
@Singleton
class ThanksClient @Inject constructor(
    @param:Named(NAMED_COMMONS_CSRF) private val csrfTokenClient: CsrfTokenClient,
    private val service: ThanksInterface
) {
    /**
     * Thanks a user for a particular revision
     * @param revisionId The revision ID the user would like to thank someone for
     * @return if thanks was successfully sent to intended recipient
     */
    fun thank(revisionId: Long): Observable<Boolean> {
        return try {
            service.thank(
                revisionId.toString(),                      // Rev
                null,                                       // Log
                csrfTokenClient.getTokenBlocking(),              // Token
                CommonsApplication.getInstance().userAgent  // Source
            ).map {
                mwThankPostResponse -> mwThankPostResponse.result?.success == 1
            }
        }
        catch (throwable: Throwable) {
            if (throwable is InvalidLoginTokenException) {
                Observable.error(throwable)
            }
            else {
                Observable.just(false)
            }
        }
    }

}
